server/src/main/java/io/crate/expression/scalar/HasTablePrivilegeFunction.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar;
+
+import static io.crate.metadata.Scalar.DETERMINISTIC_ONLY;
+
+import java.util.Collection;
+
+import io.crate.metadata.FunctionName;
+import io.crate.metadata.Functions;
+import io.crate.metadata.RelationName;
+import io.crate.metadata.Schemas;
+import io.crate.metadata.functions.Signature;
+import io.crate.metadata.pgcatalog.PgCatalogSchemaInfo;
+import io.crate.role.Permission;
+import io.crate.role.Role;
+import io.crate.role.Roles;
+import io.crate.role.Securable;
+import io.crate.types.DataTypes;
+
+public class HasTablePrivilegeFunction {
+
+    public static final FunctionName NAME = new FunctionName(PgCatalogSchemaInfo.NAME, "has_table_privilege");
+
+    public static boolean checkByTableName(Roles roles, Role user, Object table, Collection<Permission> permissions, Schemas schemas) {
+        String tableFqn = RelationName.fqnFromIndexName((String) table);
+        return checkByTablePrivileges(roles, user, tableFqn, permissions);
+    }
+
+    public static boolean checkByTableOid(Roles roles, Role user, Object table, Collection<Permission> permissions, Schemas schemas) {
+        int tableOid = (int) table;
+        RelationName relationName = schemas.getRelation(tableOid);
+        if (relationName == null) {
+            throw new IllegalArgumentException("Cannot find corresponding relation by the given oid");
+        }
+        String tableFqn = relationName.fqn();
+        return checkByTablePrivileges(roles, user, tableFqn, permissions);
+    }
+
+    private static boolean checkByTablePrivileges(Roles roles, Role user, String table, Collection<Permission> permissions) {
+        boolean result = false;
+        for (Permission permission : permissions) {
+            result |= roles.hasPrivilege(user, permission, Securable.TABLE, table);
```

Comment:
can we simply return true and break the iteration?

Reply:
I think we need it to check the user holds all permissions in the list. I will add a test for this as well.

Reply:
I've tested on postgres:
```
matriv=> select has_table_privilege('t', 'SELECT,DELETE,UPDATE');
 has_table_privilege
---------------------
 t
(1 row)

matriv=> delete from t;
ERROR:  permission denied for table t

matriv=> select * from t;
 a
---
(0 rows)

matriv=> select has_table_privilege('t', 'DELETE');
 has_table_privilege
---------------------
 f
(1 row)

```
So it seems that although the user has no delete permission but only select, the function returns true if at least on is true, so no need to do the `|=` there.

Reply:
Thanks for double checking, if so there maybe a bug in `HasSchemaPrivilegeFunction` will follow up.

Reply:
I think that for has schema it had a different usages because of implicit permissions, but yes, please check.

Reply:
> Thanks for double checking, if so there maybe a bug in HasSchemaPrivilegeFunction will follow up.

I was just confused, there is no bug, `test_user_with_DQL_permission_has_USAGE_but_not_CREATE_privilege_for_regular_schema` verifies it.